plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestStandardColumnMappings.java

```diff
@@ -0,0 +1,126 @@
+        picosOfMicro += (instant.getNano() % NANOSECONDS_PER_MICROSECOND) * PICOSECONDS_PER_NANOSECOND;
+        verifyLongTimestamp(picosOfMicro, precision);
+    }
+
+    @DataProvider
+    public Object[][] validLongTimestamps()
+    {
+        return new Object[][] {
+                {7, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_700), 0},
+                {8, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_780), 0},
+                {9, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 0},
+                {10, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 100},
+                {11, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 120},
+                {12, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 123},
+        };
+    }
+
+    @Test(dataProvider = "invalidShortTimestamps")
+    public void testVerifyInvalidShortTimestamps(int precision, LocalDateTime localDateTime, int picosOfMicro, String expectedException)
+    {
+        Instant instant = localDateTime.atZone(UTC).toInstant();
+        long epochMicros = instant.getEpochSecond() * MICROSECONDS_PER_SECOND + floorDiv(instant.getNano(), NANOSECONDS_PER_MICROSECOND);
+        int finalPicosOfMicro = picosOfMicro + (instant.getNano() % NANOSECONDS_PER_MICROSECOND) * PICOSECONDS_PER_NANOSECOND;
+        assertThatThrownBy(() -> verifyShortTimestamp(epochMicros, finalPicosOfMicro, precision)).hasMessageContaining(expectedException);
+    }
+
+    @DataProvider
+    public Object[][] invalidShortTimestamps()
+    {
+        return new Object[][] {
+                {0, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 100_000_000), 0, "Invalid value"},
+                {0, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 123, "Invalid value"},
+                {1, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 123, "Invalid value"},
+                {2, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 123, "Invalid value"},
+                {3, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 123, "Invalid value"},
+                {4, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 123, "Invalid value"},
+                {5, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 123, "Invalid value"},
+                {6, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 123, "Invalid value"},
+                {7, LocalDateTime.of(2018, 3, 25, 3, 17, 17, 123_456_789), 123, "Precision must be less"},
+                {11, LocalDateTime.of(2018, 3, 25, 3, 17, 17), 1, "Precision must be less"},
+        };
+    }
+
+    @Test(dataProvider = "invalidLongTimestamps")
+    public void testVerifyInvalidLongTimestamps(int precision, LocalDateTime localDateTime, int picosOfMicro, String expectedException)
+    {
+        Instant instant = localDateTime.atZone(UTC).toInstant();
+        int finalPicosOfMicro = picosOfMicro + (instant.getNano() % NANOSECONDS_PER_MICROSECOND) * PICOSECONDS_PER_NANOSECOND;
```

Comment:
it's counter intuitive that picos of micro are getting combined from two other values

it's a byproduct of the fact you wanted to use `LocalDateTime` for testing, but there is not advantage to it
just call the `verifyTimestamp` function with _some_ epoch-micro and some picos-of-micro. In decimal form, it looks equally readable as long as timestamp precision is concerned.

Reply:
Refactor, looks _much_ better, thanks!